Make twine Python 3 only

jdufresne opened the pull request:
I would humbly like to suggest the twine project move to be Python 3 only. As twine is a command-line tool, it is in a good position to move to Python 3 only. Other command-line-only tools have begun moving to Python 3. For example: black, mypy, pylint, pypinfo, and Sphinx. The twine command will continue to support uploading Python 2 projects to PyPI. So this should be mostly transparent to users.
    
Allows for code cleanups. By removing these workarounds, the project will be easier to maintain and reduce ongoing testing resources.
    
Python 2 is scheduled to be EOL on January 1, 2020.
    
https://www.python.org/dev/peps/pep-0373/
   
For an overview of projects that have dropped Python 2 support entirely, see:
    
https://hugovk.github.io/drop-python/2.7/

jaraco commented on twine/wheel.py:
```diff
@@ -62,10 +49,7 @@ def py_version(self):
-            x.split('/') for x in map(try_decode, names)
-            if 'METADATA' in x
-        ]
+        tuples = [x.split('/') for x in names if 'METADATA' in x]
```
This function previously would have supported bytes as input, but does no longer. Can we say with confidence that there's no Python 3 usage that might be passing bytes for names?

jdufresne replied:
Yes, I believe we can now assume this is always text. This is directly related to the changes to [test_wheel.py](https://github.com/pypa/twine/pull/437#discussion_r246014184).

The only call to `find_candidate_metadata_files()` is in `twine/wheel.py`.

https://github.com/pypa/twine/blob/5a72af19fdb27d402c7bcea41199e5eae9082fbb/twine/wheel.py#L90

The arguments to this call always come from `ZipFile.namelist()` (a few lines up in the link).

https://github.com/pypa/twine/blob/5a72af19fdb27d402c7bcea41199e5eae9082fbb/twine/wheel.py#L80

On Python 2, `ZipFile.namelist()` can return either `str` (bytes type) or `unicode`. Here is the relevant CPython code:

https://github.com/python/cpython/blob/v2.7.15/Lib/zipfile.py#L377-L381

On Python 3, `ZipFile.namelist()` always returns a `str` (text type). Here is the relevant CPython code:

https://github.com/python/cpython/blob/v3.7.2/Lib/zipfile.py#L1322-L1327

So in summary, we can now assume the value never needs to be decoded and the test can be altered to `str` (not `bytes`) to be more representative of `ZipFile`.

jaraco commented:
Aha. And now I see that in #436, you've done just as I'd have - added the Python 2 compatible changes first.

jdufresne commented on twine/commands/check.py:
```diff
@@ -11,18 +11,11 @@
-    from StringIO import StringIO
-except ImportError:
-    from _io import StringIO
+from io import StringIO
```
:+1: Done in #444

jdufresne commented on tests/test_wheel.py:
```diff
@@ -32,11 +30,11 @@ def test_version_parsing(example_wheel):
-        b'package/METADATA.txt',
-        b'package/METADATA.json',
-        b'package/METADATA',
+        'package/lib/__init__.py',
```
See https://github.com/pypa/twine/pull/437#discussion_r246244022 for an explanation why this change must wait for dropping Python 2.

jdufresne commented on setup.cfg:
```diff
@@ -1,6 +1,3 @@
-[bdist_wheel]
```
According to the [wheel documentation](https://wheel.readthedocs.io/en/stable/user_guide.html), a universal wheel is defined as:

> If your project contains no C extensions and is expected to work on both Python 2 and 3, you will want to tell wheel to produce universal wheels by adding this to your setup.cfg file:

With this change, Twine no longer works on **both** Python 2 and 3, so I believe it is incorrect to declare the wheel as universal. Just being pure Python is not enough to be declared universal.

For an example from another project, Django dropped universal when dropping Python 2. See commit:

https://github.com/django/django/commit/1c466994d9729a8902d34e277a3a685bef2a08e1

jdufresne replied:
One idea would be to add a check:

```py
if sys.version_info < (3,4):
    raies ...
```
 This would prevent someone from creating a Python2 wheel. WDYT?

jdufresne replied:
I made a suggestion with the setuptools project that this could be handled more centrally by `python_requires`: https://github.com/pypa/setuptools/issues/1633

jaraco replied:
I’d try to avoid imperative code in setup.py. I do agree that if setuptools could derive this information from other signals that would be better. Happy to keep this change as-is.

jdufresne commented:
Thanks very much for the review @jaraco! I have rebased on the latest master so this PR no longer includes Python 2 style changes and extracted the `io.StringIO` change as suggested. I have responded to the comments that I believe should remains as-is. Please let me know if you have additional feedback or suggestions. :slightly_smiling_face:

sigmavirus24 commented:
I want to hold off on this until we can cut a 2.0 and I want to cut a different release today

jdufresne commented:
I have rebased to resolve merge conflicts. I also adjusted the tox environment "release" with `basepython = python3` to help further mitigate the concern with mistakenly releasing a Python 2 wheel.

> I want to hold off on this until we can cut a 2.0 and I want to cut a different release today

Sounds good to me.

jaraco commented:
I'm ready to accept this. @theacodes do you have any objections?

sigmavirus24 commented:
Isn't py3.4 EOL? Should we raise the minimum supported version to 3.5 or 3.6?

jdufresne commented:
Yeah, I agree. I've removed Python 3.4 support in the latest revision.

bhrutledge commented:
I'm working on type annotations for #231. If twine required 3.6+, we could use the [new syntax for variable annotations](https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#variables), instead of type comments. I don't know what other considerations there might be for dropping 3.5.

theacodes commented:
No objections

On Tue, Jun 11, 2019, 6:06 PM Brian Rutledge <notifications@github.com>
wrote:

> I'm working on type annotations for #231
> <https://github.com/pypa/twine/issues/231>. If twine required 3.6+, we
> could use the new syntax for variable annotations
> <https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#variables>,
> instead of type comments. I don't know what other considerations there
> might be for dropping 3.5.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/twine/pull/437?email_source=notifications&email_token=AAB5I434VEMD3QTLDNOZDODP2BDYVA5CNFSM4GJG4CDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXO6CFI#issuecomment-501080341>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB5I43KBYVCZQE4JEQKWJTP2BDYVANCNFSM4GJG4CDA>
> .
>

jdufresne commented:
I have added a commit to remove Python 3.5 support.

bhrutledge commented:
@jaraco Any more thoughts about merging this? Among other things, it'd be nice to use as a baseline for adding type annotations.

bhrutledge commented:
@pypa/twine-maintainers How do y'all feel about merging this? It makes Twine require Python 3.6+, and drops support for PyPy.

di commented:
@sigmavirus24 said above:

> I want to hold off on this until we can cut a 2.0 and I want to cut a different release today

I agree that this belongs in a 2.x release, and I think we should continue merging and releasing 1.x until the EOL, so at the earliest I'd support merging & releasing this on Jan 1, 2020.

jaraco commented:
My preference would be to merge this and cut a release, but maintain a 1.x branch for bug fixes or other contributions deemed worthy of the effort.

sigmavirus24 commented:
Right. I think at that time there was a "We'd like to rush something out" release and now is a perfectly fine time to release 2.x. We can branch a 1.x branch as @jaraco says and backport to that until Jan 1, 2020 as necessary.

jaraco commented:
I was actually hoping to roll this out today, but it's blocked on #468.

jaraco commented:
I've created a `maint/1.x` branch for maintaining changes for 1.x releases as needed.

di commented on setup.py:
```diff
@@ -68,7 +64,7 @@
     },
 
-    python_requires=">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*",
+    python_requires=">=3.6",
```
I'm sorry I didn't get to comment before this was merged, but while I agree with everything else in this PR, I think this is overly restrictive. There is a difference here in saying what we will support (via classifiers) vs. what users are able to install. 

Unless we know for sure that `twine` will not work with `<=3.5`, we are excluding users from using `twine` and newer features without much reason.

I think this should probably be `>=3.4` and we should continue running CI against 3.4 and 3.5 as "allowed failures" and make this more restrictive once/if they start failing.

jaraco replied:
The patch does require Python 3.6 due to use of f-strings. Removing the `python_requires` and running `tox -e py35` yields this result:

```
ImportError while loading conftest '/Users/jaraco/code/public/pypa/twine/tests/conftest.py'.
tests/conftest.py:5: in <module>
    from twine import settings
twine/settings.py:16: in <module>
    from twine import repository
E     File "/Users/jaraco/code/public/pypa/twine/twine/repository.py", line 111
E       print(f"Registering {package.basefilename}")
E                                                 ^
E   SyntaxError: invalid syntax
ERROR: InvocationError for command /Users/jaraco/code/public/pypa/twine/.tox/py35/bin/coverage run --source twine -m pytest tests (exited with code 4)
```

I'm personally a big fan of f-strings, so I'm slightly disinclined to backport to Python 3.5. I'm more strongly resistant to backporting to Python 3.4, given that it's reached EOL (and twine 1.x still supports it).

In a61a4be, I've employed future-fstrings to keep f-strings, but tests still fail due to a missing hash implementation:

```
ImportError while importing test module '/Users/jaraco/code/public/pypa/twine/tests/test_check.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_check.py:16: in <module>
    from twine.commands import check
twine/commands/check.py:23: in <module>
    from twine.package import PackageFile
twine/package.py:21: in <module>
    from hashlib import blake2b
E   ImportError: cannot import name 'blake2b'
________________________________________
```

I haven't gone farther to determine what other Python 3.6 features might be present.

In light of those concerns, should we invest the energy to continue to backport to 3.4 or 3.5?

di replied:
I see now that we've undone some compatability code in this PR.

I think generally our policy should be not to do this unless we really need to use some feature or library.

I'm not sure if it's worth adding compatability back at this point: the goal should be for us to do less work, not more.

bhrutledge replied:
I realize this is a small consideration, and squarely in the developer experience realm, but I noted in https://github.com/pypa/twine/pull/437#issuecomment-501080341 that requiring 3.6+ would let us use [type annotations on variables](https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#variables) (instead of type comments).

jaraco replied:
Given that `python_requires>=3.6` will prevent this version from being installed on Python 3.5 and earlier (pip 9+), _and_ that Twine 1.x will be installed instead, I'm very much in favor of moving forward. Thanks for raising this concern, though. I learned some things in the investigation that I'd missed before.
